### ISSUE TITLE
Nested schema type causing ObjectParameterError: Parameter "obj" to Document() must be an object, got "..." (type string) when `runValidators` is `true`

### ISSUE THREAD

**Issue by fredkilbourn** [Automattic/mongoose#13977]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Mongoose version

7.6.2

### Node.js version

18.16.0

### MongoDB server version

6.0.0

### Description

I have a model with nested schema type, starting in version 7.6.2 it is throwing an error.  Versions 7.6.1 and before are not.

### Steps to Reproduce

Schema definition (using a number of helper functions which are not included for brevity's sake):
```javascript
const Ps2Outfit = MongoDBService.model( "Ps2Outfit",
{
        _id:            { type: String, required: true, validate: Ps2Validate.outfitId },
        name:           { type: String, required: true, validate: Ps2Validate.outfitName },
        alias:          { type: String, default: null, required: false, validate: a => a === null || Ps2Validate.outfitAlias( a ) },
        member_count:   { type: Number, required: true, min: 1, validate: Number.isInteger },
        members:
        {
                type: Map,
                of:
                {
                        type: String,
                        validate:
                        {
                                validator: ( value, validatorProperties ) =>
                                {
                                        const key = validatorProperties.path.replace( /^members\./, "" );
                                        return key === value && Ps2Validate.characterId( key );
                                },
                                propsParameter: true    //causes validatorProperties argument to be passed
                        }
                },
                required: true,
                ref: "Ps2Character",
                validate: function( m )
                {
                        if( m instanceof Map === false )
                                return false;

                        //update validators won't have access to this, so do best to handle both cases
                        if( this )
                                return this.get( "member_count" ) === m.size
                        else
                                return m.size >= 1;
                }
        },
        ranks:
        {
                type: MongoDBService.schema(
                {
                        "1":    { type: String, required: true, default: "Leader", validate: Ps2Validate.outfitRankName },
                        "2":    { type: String, required: true, default: "Officer", validate: Ps2Validate.outfitRankName },
                        "3":    { type: String, required: true, default: "Member", validate: Ps2Validate.outfitRankName },
                        "4":    { type: String, required: true, default: "Private", validate: Ps2Validate.outfitRankName },
                        "5":    { type: String, required: true, default: "Enlisted", validate: Ps2Validate.outfitRankName },
                        "6":    { type: String, required: true, default: "Recruit", validate: Ps2Validate.outfitRankName },
                        "7":    { type: String, required: true, default: "Considered", validate: Ps2Validate.outfitRankName },
                        "8":    { type: String, required: true, default: "Applicant", validate: Ps2Validate.outfitRankName },
                } ),
                required: true,
                default: {}     //required to init the above default values if nothing provided
        }
} )
```
Example upsert code:
```javascript
const outfit_object = {
  _id: '37509488620601583',
  name: 'The Vindicators',
  alias: 'V',
  member_count: 78,
  members: Map(78) {
    '5428010618019180225' => '5428010618019180225',
    '5428010618020700593' => '5428010618020700593',
    '5428010618020705601' => '5428010618020705601',
    '5428010618034974001' => '5428010618034974001',
    '5428010618035970737' => '5428010618035970737',
    '5428010618035987073' => '5428010618035987073',
    '5428010618038275857' => '5428010618038275857',
    '5428010618038458129' => '5428010618038458129',
    '5428010618040536401' => '5428010618040536401',
    '5428010618043590449' => '5428010618043590449',
    '5428010917270450129' => '5428010917270450129',
    '5428011263292885809' => '5428011263292885809',
    '5428011263352489137' => '5428011263352489137',
    '5428011263403915889' => '5428011263403915889',
    '5428011263416222065' => '5428011263416222065',
    '5428013610393813217' => '5428013610393813217',
    '5428013610423625361' => '5428013610423625361',
    '5428013610428730273' => '5428013610428730273',
    '5428013610474366833' => '5428013610474366833',
    '5428013610481107857' => '5428013610481107857',
    '5428013610486769137' => '5428013610486769137',
    '5428016813491901761' => '5428016813491901761',
    '5428019223849228097' => '5428019223849228097',
    '5428034441445470273' => '5428034441445470273',
    '5428041429989239025' => '5428041429989239025',
    '5428041429994590065' => '5428041429994590065',
    '5428047126265519233' => '5428047126265519233',
    '5428047126274020257' => '5428047126274020257',
    '5428052273451385265' => '5428052273451385265',
    '5428052273454630833' => '5428052273454630833',
    '5428061686722498417' => '5428061686722498417',
    '5428061686734840241' => '5428061686734840241',
    '5428066778516885553' => '5428066778516885553',
    '5428072203495682273' => '5428072203495682273',
    '5428077659950345681' => '5428077659950345681',
    '5428095401453020865' => '5428095401453020865',
    '5428095401453349649' => '5428095401453349649',
    '5428117870045771073' => '5428117870045771073',
    '5428130184558856769' => '5428130184558856769',
    '5428147970869644113' => '5428147970869644113',
    '5428153774128421873' => '5428153774128421873',
    '5428163555326488817' => '5428163555326488817',
    '5428168624804679745' => '5428168624804679745',
    '5428168624828018177' => '5428168624828018177',
    '5428201940314319489' => '5428201940314319489',
    '5428222237686032481' => '5428222237686032481',
    '5428224050372094401' => '5428224050372094401',
    '5428232026855027601' => '5428232026855027601',
    '5428280608551693793' => '5428280608551693793',
    '5428285306556624593' => '5428285306556624593',
    '5428297992008446401' => '5428297992008446401',
    '5428335660029406737' => '5428335660029406737',
    '5428359100720151969' => '5428359100720151969',
    '5428437870208430369' => '5428437870208430369',
    '5428491127575260769' => '5428491127575260769',
    '5428564342235979697' => '5428564342235979697',
    '5428587263587520369' => '5428587263587520369',
    '5428602376701849729' => '5428602376701849729',
    '5428624844701326065' => '5428624844701326065',
    '5428653158002731265' => '5428653158002731265',
    '5428660720835383521' => '5428660720835383521',
    '5428721773502237937' => '5428721773502237937',
    '5428754945603497377' => '5428754945603497377',
    '5428754945603876481' => '5428754945603876481',
    '5428812948092007025' => '5428812948092007025',
    '5428812948101240033' => '5428812948101240033',
    '5428861139969144977' => '5428861139969144977',
    '5428864149645660081' => '5428864149645660081',
    '5428906429081985329' => '5428906429081985329',
    '5428995209017635633' => '5428995209017635633',
    '5428998138859160785' => '5428998138859160785',
    '5429033621897071553' => '5429033621897071553',
    '5429088717264060097' => '5429088717264060097',
    '5429091716540082433' => '5429091716540082433',
    '5429137261034830257' => '5429137261034830257',
    '5429170278625327633' => '5429170278625327633',
    '5429286928893140401' => '5429286928893140401',
    '5429442030439944769' => '5429442030439944769'
  },
  ranks: {
    '1': 'Bishop',
    '2': 'Priest',
    '3': 'Fanatic',
    '4': 'Zealot',
    '5': 'Chaplain',
    '6': 'Deacon',
    '7': 'Acolyte',
    '8': 'Brother'
  }
};
await Ps2Outfit.updateOne( { _id: outfit_object._id }, outfit_object, { upsert: true } );
```
Resulting error:
```
ObjectParameterError: Parameter "obj" to Document() must be an object, got "Bishop" (type string)
    at SingleNested.Document (node_modules/mongoose/lib/document.js:111:11)
    at SingleNested.Subdocument (node_modules/mongoose/lib/types/subdocument.j
    at new SingleNested (node_modules/mongoose/lib/schema/SubdocumentPath.js:8
    at SubdocumentPath.doValidate (node_modules/mongoose/lib/schema/Subdocumen
    at node_modules/mongoose/lib/helpers/updateValidators.js:151:22
    at module.exports (node_modules/mongoose/lib/helpers/updateValidators.js:2
    at node_modules/mongoose/lib/query.js:3930:7
    at new Promise (<anonymous>)
    at model.Query.validate (node_modules/mongoose/lib/query.js:3929:11)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
```

### Expected Behavior

In versions 7.6.1 and before for years this code functioned fine and the upsert completed without error, this is what I would expect to happen.

**Comment by IslandRhythms** [comment]
Please modify the script below to demonstrate your issue.

```javascript
const mongoose = require('mongoose');

const testMap = new Map();

const testSchema = new mongoose.Schema({
  name: String,
  alias: String,
  member_count: Number,
  members: {
    type: Map,
    validate: function (m) {
      if (m instanceof Map === false) {
        return false;
      }
      if (this) {
        return this.get("member_count") === m.size
      } else {
        return m.size >= 1;
      }
    },
    of: {
      type: String,
      validate: {
        validator: (value, validatorProperties) => {
          const key = validatorProperties.path.replace( /^members\./, "" );
          return key === value;
        },
        propsParameter: true
      }
    }
  },
  ranks: {
    type: new mongoose.Schema({
      "1": String,
      "2": String,
      "3": String,
      "4": String,
      "5": String,
      "6": String,
      "7": String,
      "8": String
    }),
    required: true,
    default: {}
  }
});

const Test = mongoose.model('Test', testSchema);

testMap.set('741025896303','369852014741');
testMap.set('852369014741', '789654123210');
const outfit_object = {
  _id: '375094886206015834569321',
  name: 'The Vindicators',
  alias: 'V',
  member_count: 78,
  members: testMap,
  ranks: {
    '1': 'Bishop',
    '2': 'Priest',
    '3': 'Fanatic',
    '4': 'Zealot',
    '5': 'Chaplain',
    '6': 'Deacon',
    '7': 'Acolyte',
    '8': 'Brother'
  }
};

async function run() {
  await mongoose.connect('mongodb://localhost:27017');
  await mongoose.connection.dropDatabase();

  await Test.updateOne({}, outfit_object, { upsert: true });
  console.log('done');
}

run();
```

**Comment by fredkilbourn** [comment]
Hi @IslandRhythms,

I consolidated all my code based on your example:
```javascript
import mongoose from 'mongoose';

mongoose.set( 'runValidators', true );

//define mongoose connect parameters
const hosts =
[
        'REDACTED',
        'REDACTED',
        'REDACTED',
];
const url = `mongodb://${hosts.join()}`;
const options =
{
        //config based options
        tls:                            true,
        dbName:                         'REDACTED',
        authSource:                     'REDACTED',
        auth:
        {
                username:               'REDACTED',
                password:               'REDACTED'
        },

        //mongoose options
        autoIndex:                      true,
        bufferCommands:                 false,
        serverSelectionTimeoutMS:       5000,

        //mongodb options
        appname:                        'REDACTED',
        authMechanism:                  'SCRAM-SHA-256'
};
options.tlsCAFile = 'REDACTED';
options.replicaSet = 'REDACTED';

//connect
await mongoose.connect( url, options );

class Ps2Validate
{
        static outfitId( id )
        {
                return is.nonEmptyString( id ) &&
                        id.match( /^[0-9]{17}$/ ) !== null;
        }

        static outfitRankName( name )
        {
                //longest 20 based on max() in the current database
                //After seeing almost any kind of character, we only check length
                return is.nonEmptyString( name ) && name.length <= 20;
        }
};

const Ps2OutfitRankSchema = mongoose.Schema(
{
        '1':    { type: String, required: true, default: 'Leader', validate: Ps2Validate.outfitRankName },
        '2':    { type: String, required: true, default: 'Officer', validate: Ps2Validate.outfitRankName },
        '3':    { type: String, required: true, default: 'Member', validate: Ps2Validate.outfitRankName },
        '4':    { type: String, required: true, default: 'Private', validate: Ps2Validate.outfitRankName },
        '5':    { type: String, required: true, default: 'Enlisted', validate: Ps2Validate.outfitRankName },
        '6':    { type: String, required: true, default: 'Recruit', validate: Ps2Validate.outfitRankName },
        '7':    { type: String, required: true, default: 'Considered', validate: Ps2Validate.outfitRankName },
        '8':    { type: String, required: true, default: 'Applicant', validate: Ps2Validate.outfitRankName },
},
{
        id:             false,
        _id:            false,
        strict:         'throw',
        strictQuery:    'throw',
        toObject:       { virtuals: true }
} );

const Ps2OutfitSchema = mongoose.Schema(
{
        _id:            { type: String, required: true, validate: Ps2Validate.outfitId },
        ranks:
        {
                type: Ps2OutfitRankSchema,
                required: true,
                default: {}     //required to init the above default values if nothing provided
        }
},
{
        id:             false,
        _id:            true,
        strict:         'throw',
        strictQuery:    'throw',
        toObject:       { virtuals: true },
        timestamps:     true
} );

const Ps2OutfitTest = mongoose.model( 'Ps2OutfitTest', Ps2OutfitSchema, 'Ps2OutfitTest' );

const outfit_object = {
  _id: '37509488620601583',
  ranks: {
    '1': 'Bishop',
    '2': 'Priest',
    '3': 'Fanatic',
    '4': 'Zealot',
    '5': 'Chaplain',
    '6': 'Deacon',
    '7': 'Acolyte',
    '8': 'Brother'
  }
};

await Ps2OutfitTest.updateOne( { _id: outfit_object._id }, outfit_object, { upsert: true } );

await mongoose.disconnect();
```

Through trial and error, I found the key to throwing the exception was this line:
```javascript
mongoose.set( 'runValidators', true );
```
When set to true, the error happens, and the error doesn't happen if set to false.

I would still like to use runValidators: true, and I think this is still a bug?

I hope this helps you repro more easily.

**Comment by hassan-sh** [comment]
👋 Hi there! I encountered the exact same issue when trying to increment a subdocument and wanted to provide some additional context that might help.

**Model Definition:**
```
const errorResponsesSchema = mongoose.Schema(
  {
    '403': Number,
    '404': Number,
    '410': Number,
    'otherCode': Number,
  },
  { _id: false }
);

const reportSchema = new mongoose.Schema(
  {
    project: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "project",
      required: true
    },
    title: {
      type: String,
      required: true
    },
    body: {
      type: String,
      required: true
    },
    errorResponses:  errorResponsesSchema
  },
  { timestamps: true }
);
```

**Update Operation:**
```
await Notification.findByIdAndUpdate(report_id, {$inc: { 'errorResponses.403': 1 } })
```

**Error Message:**
```
{
    "stringValue": "\"1\"",
    "valueType": "number",
    "kind": "Embedded",
    "value": 1,
    "path": "errorResponses",
    "reason": {},
    "name": "CastError",
    "message": "Cast to Embedded failed for value \"1\" (type number) at path \"errorResponses\" because of \"ObjectParameterError\""
}
```

I observed this error starting from version 7.6.2 up to 8.0.0. Interestingly, everything worked as expected on version 7.6.1 and earlier versions. It seems like the $inc operator isn't behaving as expected with subdocument fields in the newer versions.

I noticed the mention about the setting:
```
mongoose.set( 'runValidators', true );
```
 I've tried setting it to both true and false, but the error persisted in my case regardless of the setting's value.

I hope this additional information helps in diagnosing and fixing the issue

**Comment by vkarpov15** [comment]
We merged the fix to master branch as well, so fix will be in 8.0.2
